### DISCUSSION TITLE
Milvus response time for vector search fluctuatesrapidly

### DISCUSSION THREAD

**Discussion by Pavan29082003** [milvus-io/milvus#37051] [Q&A and General discussion]
I am experiencing significant fluctuations in query times while running Milvus on an EC2 instance using Docker. Initially, query times are fast (around 1000ms), but they occasionally spike to over 9s, and this happens inconsistently during normal operations. The issue becomes more pronounced as the data volume increases.

**Comment by Pavan29082003** [comment]
I am hosting milvus on ec2 using this docker file 

ersion: '3.5'

services:
  etcd:
    container_name: milvus-etcdd
    image: quay.io/coreos/etcd:v3.5.5
    environment:
      - ETCD_AUTO_COMPACTION_MODE=revision
      - ETCD_AUTO_COMPACTION_RETENTION=1000
      - ETCD_QUOTA_BACKEND_BYTES=4294967296
      - ETCD_SNAPSHOT_COUNT=50000
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/etcd:/etcd
    command: etcd -advertise-client-urls=http://127.0.0.1:2379 -listen-client-urls http://0.0.0.0:2379 --data-dir /etcd

  minio:
    container_name: milvus-minio
    image: minio/minio:RELEASE.2023-03-20T20-16-18Z
    environment:
      MINIO_ACCESS_KEY: minioadmin
      MINIO_SECRET_KEY: minioadmin
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/minio:/minio_data
    command: minio server /minio_data
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
      interval: 30s
      timeout: 20s
      retries: 3
  standalone:
    container_name: milvus-standalone
    image: milvusdb/milvus:v2.4.0
    command: ["milvus", "run", "standalone"]
    environment:
      ETCD_ENDPOINTS: etcd:2379
      MINIO_ADDRESS: minio:9000
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/milvus:/var/lib/milvus
    ports:
      - "19530:19530"
      - "9091:9091"
    depends_on:
      - "etcd"
      - "minio"
  attu:
    container_name: attu
    image: zilliz/attu:v2.2.6
    environment:
      MILVUS_URL: milvus-standalone:19530
    ports:
      - "8000:3000"
    depends_on:
      - "standalone"

networks:
  default:
    name: milvus

**Comment by Pavan29082003** [comment]
![Screenshot (25)](https://github.com/user-attachments/assets/8f40cdf7-e8ec-4135-a820-edc028bdf919)

This the memory usage of all containers that are in my ec2.There is no fluctuation in this data while I am querying from milvus

**Comment by yhmo** [comment]
It sounds like your use case is to continually insert data and query.
Some questions:
1. Which version of the milvus?
2. How many entities are in the collection? Could you show me the collection's schema?
3. Averagely how many entities are inserted into the collection per second?
4. Which interface fluctuates? collection.search() or collection.query()? Did you set "consistency_level"? Did you set "output_fields"?

**Reply by Pavan29082003** [reply]
milvus_version - 2.4.0
 output fields -   [ * ]
I am performing collection.search() on three collections using multi-threading at the same time.The entities in the collection are as follows: 29,775, 84,467, 57,706.
I am uploading the collection schema for one collection which is same for all collection I am querying 
![Screenshot (26)](https://github.com/user-attachments/assets/969a5497-c9a0-4823-af4d-7eb4ed5516d8)

**Reply by Pavan29082003** [reply]
Also I dont continously insert data.All the data is inserted I am only trying to find similar vector based on a query

**Reply by yhmo** [reply]
Which index type?

How many vectors are input to collection.search() to search per time? What value of the "limit"
```
res = collection.search(
     data=vectors,  # how many vectors ?
     anns_field="vector",
     limit=limit,   # what value for this limit?
     param={},
     output_fields=["*"],
 )
```

The "boay_content" is a json type field, could be a long-length content?

**Reply by Pavan29082003** [reply]
Index type - IVF_FLAT
Input Vector - 1 [384 Dimensions]
limit - 100
Yes the body_content can be long-length content

I am also posting the python code i am using to query content

import threading
from concurrent.futures import ThreadPoolExecutor
from sentence_transformers import SentenceTransformer
from pymilvus import MilvusClient, Collection, connections, MilvusException
from src.settings import settings
import time

client = MilvusClient(uri="http://" + str(settings.ip) + ":19530")
connections.connect(host=str(settings.ip), port="19530")

vector_data_pmc = Collection(name="vector_data_pmc")
vector_data_biorxiv = Collection(name="vector_data_biorxiv")
vector_data_plos = Collection(name="vector_data_plos")

sbert_model = SentenceTransformer('all-MiniLM-L6-v2')

def search_milvus(collection, query_embedding):
    return collection.search(
        param={"metric_type": "COSINE", "params": {}},
        data=query_embedding,
        anns_field="vector_data",
        limit=100,
        output_fields=["*"]
    )

def hit_to_dict(hit):
    hit_dict = {'similarity_score': hit.score}
    field_names = ['pmid', 'pmc', 'biorxiv_id', 'plos_id', 'source', 'body_content', 'abstract_content', 
                   'publication_date', 'publication_type', 'figures', 'article_title']
    
    for field_name in field_names:
        try:
            value = getattr(hit, field_name, None)
            if not isinstance(value, (str, int, list, dict)):
                hit_dict[field_name] = list(value)
            else:
                hit_dict[field_name] = value
        except MilvusException:
            continue
    
    return hit_dict

def get_data(query):
    query_embedding = sbert_model.encode([query])
    start_time = time.time()
    with ThreadPoolExecutor(max_workers=3) as executor:
        search_results = list(executor.map(
            lambda collection: search_milvus(collection, query_embedding),
            [vector_data_pmc, vector_data_biorxiv, vector_data_plos]
        ))
    end_time = time.time()
    print("Time for vector search:" ,end_time - start_time)
    res_pmc, res_biorxiv, res_plos = search_results
    relevant_articles = []

    for hits_pmc, hits_biorxiv, hits_plos in zip(res_pmc, res_biorxiv, res_plos):
        for hit_pmc, hit_biorxiv, hit_plos in zip(hits_pmc, hits_biorxiv, hits_plos):
            relevant_articles.extend([
                hit_to_dict(hit_pmc),
                hit_to_dict(hit_biorxiv),
                hit_to_dict(hit_plos)
            ])

    articles = sorted(relevant_articles, key=lambda x: x['similarity_score'], reverse=True)

    for article in articles:
        article['similarity_score'] = ((article['similarity_score'] + 1) / 2) * 100

    return articles


I am calculating the time it is taking for query using time module in python.Here are the timings

Time for vector search: 3.392845392227173
INFO:     127.0.0.1:63346 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for vector search: 6.044903516769409
INFO:     127.0.0.1:63576 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for vector search: 1.4971187114715576
INFO:     127.0.0.1:63583 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for vector search: 1.9726142883300781
INFO:     127.0.0.1:63652 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for vector search: 2.6184475421905518
INFO:     127.0.0.1:63771 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for vector search: 14.678842067718506
INFO:     127.0.0.1:63780 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for vector search: 24.950464010238647
INFO:     127.0.0.1:63797 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for vector search: 15.042075634002686

**Reply by yhmo** [reply]
Modify the search_milvus method to be:
```
def search_milvus(collection, query_embedding):
    start_time = time.time()
    res = collection.search(
        param={"metric_type": "COSINE", "params": {}},
        data=query_embedding,
        anns_field="vector_data",
        limit=100,
        output_fields=["*"]
    )
    end_time = time.time()
    print("Time for collection.search:", end_time - start_time)
    return res
```
And try running the script, how about the internal time vs the outside time?

**Reply by yhmo** [reply]
Set "output_fields=["*"]" to output_fields=["vector_data"] and observe the time change.

**Reply by Pavan29082003** [reply]
These are the logs showing time for querying using the script you have provided

INFO:     Started server process [9876]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
Time for collection.search: 1.9194810390472412
Time for collection.search: 1.950124979019165
Time for collection.search: 1.980851650238037
INFO:     127.0.0.1:64996 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 1.090721845626831
Time for collection.search: 1.5931367874145508
Time for collection.search: 1.6121430397033691
INFO:     127.0.0.1:64996 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 1.8397555351257324
Time for collection.search: 3.2410874366760254
Time for collection.search: 3.2723777294158936
INFO:     127.0.0.1:65002 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 7.281263589859009
Time for collection.search: 7.44113302230835
Time for collection.search: 7.502291440963745
INFO:     127.0.0.1:65004 - "POST /core_search/query/ HTTP/1.1" 200 OK

These are the logs when only ['vector_data'] is placed in output field

NFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.15502667427062988
Time for collection.search: 0.3547182083129883
Time for collection.search: 0.353226900100708
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.16137123107910156
Time for collection.search: 0.3442420959472656
Time for collection.search: 0.5552873611450195
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.2324669361114502
Time for collection.search: 0.4732999801635742
Time for collection.search: 0.6947953701019287
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.20538067817687988
Time for collection.search: 0.3443610668182373
Time for collection.search: 0.4840841293334961
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.18820977210998535
Time for collection.search: 0.2833540439605713
Time for collection.search: 0.38144636154174805
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.2940943241119385
Time for collection.search: 0.3334677219390869
Time for collection.search: 0.3334643840789795
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.10228633880615234
Time for collection.search: 0.13680481910705566
Time for collection.search: 0.15134072303771973
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.09579181671142578
Time for collection.search: 0.12880158424377441
Time for collection.search: 0.14531326293945312
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.12139558792114258
Time for collection.search: 0.15093231201171875
Time for collection.search: 0.1741940975189209
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.16252994537353516
Time for collection.search: 0.17052984237670898
Time for collection.search: 0.17809343338012695
INFO:     127.0.0.1:65045 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.20920133590698242
Time for collection.search: 0.2134077548980713
Time for collection.search: 0.4113795757293701


I have previously tried this approach which is gettting only id from collection.seach() and querying the three collections using collection.get() ids to get all the fields.In this approach the vector search was happening quickly but collection.get() was taking so much time.

**Reply by yhmo** [reply]
I guess the bottleneck is in retrieving the long-length content such as "body_content" from S3.
You can try mmap:

```
vector_data_pmc.release()
vector_data_pmc.set_properties({'mmap.enabled': True})
vector_data_pmc.load()

vector_data_biorxiv.release()
vector_data_biorxiv.set_properties({'mmap.enabled': True})
vector_data_biorxiv.load()

vector_data_plos.release()
vector_data_plos.set_properties({'mmap.enabled': True})
vector_data_plos.load()
```

**Reply by yhmo** [reply]
To confirm which field is the bottleneck, you can remove the "body_content" from the `output_fields` and try again.
```
res = collection.search(
        ......
        output_fields=["article_title", "pmid", "pmc", "abstract_content", "figures", "vector_data", "publication_type", "publication_date"]
    )
```

**Reply by Pavan29082003** [reply]
But I need body_content.It is the main field of my collection

**Reply by Pavan29082003** [reply]
These are the logs after tryng mmap

Time for collection.search: 7.655706882476807
Time for collection.search: 8.116857290267944
Time for collection.search: 8.207449674606323
INFO:     127.0.0.1:65229 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 7.870658874511719
Time for collection.search: 8.632124423980713
Time for collection.search: 8.663057804107666
INFO:     127.0.0.1:65229 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 5.304045915603638
Time for collection.search: 5.913760423660278
Time for collection.search: 5.974615097045898
INFO:     127.0.0.1:65238 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 4.697610378265381
Time for collection.search: 5.139774322509766
Time for collection.search: 5.199482679367065
INFO:     127.0.0.1:65247 - "POST /core_search/query/ HTTP/1.1" 200 OK

**Reply by yhmo** [reply]
Ok, mmap is slower. Let's forget the mmap.

```
vector_data_pmc.release()
vector_data_pmc.set_properties({'mmap.enabled': False})
vector_data_pmc.load()

vector_data_biorxiv.release()
vector_data_biorxiv.set_properties({'mmap.enabled': False})
vector_data_biorxiv.load()

vector_data_plos.release()
vector_data_plos.set_properties({'mmap.enabled': False})
vector_data_plos.load()
```

**Reply by yhmo** [reply]
The performance issue is about the long-length JSON content.
Could you show me an example of the "body_content"?

**Reply by Pavan29082003** [reply]
Would like to connect in a call only if you are comfortable

**Reply by Pavan29082003** [reply]
This is a sample record in the collection
[milvus_query_result (13).csv](https://github.com/user-attachments/files/17474089/milvus_query_result.13.csv)

**Reply by Pavan29082003** [reply]
Sometimes the query is being done very quickly even without using any mmap or anything

INFO:     127.0.0.1:65446 - "POST /core_search/query/ HTTP/1.1" 200 OK
Time for collection.search: 0.6692612171173096
Time for collection.search: 0.6932282447814941
Time for collection.search: 0.7056913375854492

**Reply by Pavan29082003** [reply]
Here output fields is also set to ["*"]

**Reply by yhmo** [reply]
The time cost mainly depends on the length of the JSON content. Maybe the "body_content" of some rows is simple then the query is faster.

**Reply by Pavan29082003** [reply]
I have been using same input query vector throughout the day.So the results that are being returned are all same each and every time

**Reply by yhmo** [reply]
A workaround you can try:
1. create another three collections, the same schema, except the "body_content" is a Varchar field
2. in client side, use json.dumps() to convert the body_content to a str, and insert into the new collections.
3. query in the same way, output_fields="*"
4. in client side, use json.loads() to convert the body_content string to a json object.

I am not sure how much faster is the workaround, I need to do some local tests to verify this performance issue.

**Reply by Pavan29082003** [reply]
Removing body_content from output fields made a lot of differnece.Now I am able to query in just 500 ms.Thank you so much for identifyng the problem

**Reply by yhmo** [reply]
@Pavan29082003 
I just used your script to test in my local. 
100000 entities for each collection with the "body_content" you provided. 
The average time cost is no more than 100ms per search. I guess the bottleneck is in your network bandwidth.
